larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/ShowerESTAREnergy_tool.cc

```diff
@@ -0,0 +1,293 @@
+  int CalculateElement(const art::Ptr<recob::PFParticle>& pfparticle,
+  art::Event& Event,
+  reco::shower::ShowerElementHolder& ShowerElementHolder
+  ) override;
+
+  private:
+    
+  double CalculateEnergy(const detinfo::DetectorClocksData& clockData,
+           const detinfo::DetectorPropertiesData& detProp, 
+           const std::vector<art::Ptr<recob::Hit> >& hits,
+           const geo::PlaneID::PlaneID_t plane);
+
+  art::InputTag fPFParticleLabel;
+  int fVerbose;
+
+  //Services
+  art::ServiceHandle<geo::Geometry> fGeom;
+  calo::CalorimetryAlg              fCalorimetryAlg;    
+
+  //fcl params 
+  std::string fShowerEnergyOutputLabel;
+  std::string fShowerBestPlaneOutputLabel; 
+  bool fSCECorrectEField;
+
+  double nominal_Efield = 0;         // Nominal E-field in the detector
+  double localEfield = 0;            // Local E-field at the geometric centre of the shower 
+  double localEfield_cweighted = 0;  // Local E-field at the charge weighted centre of the shower
+
+  std::vector<std::tuple<int, double>> sp_hits_tuple; // tuple holding the key and local E-field for all the sp hits 
+
+  TGraph2D *t_estar;
+
+  }; // class
+
+  ShowerESTAREnergy::ShowerESTAREnergy(const fhicl::ParameterSet& pset) :
+    IShowerTool(pset.get<fhicl::ParameterSet>("BaseTools")),
+    fPFParticleLabel(pset.get<art::InputTag>("PFParticleLabel")),
+    fVerbose(pset.get<int>("Verbose")),
+    fCalorimetryAlg(pset.get<fhicl::ParameterSet>("CalorimetryAlg")),
+    fShowerEnergyOutputLabel(pset.get<std::string>("ShowerEnergyOutputLabel")),
+    fShowerBestPlaneOutputLabel(pset.get<std::string>("ShowerBestPlaneOutputLabel")),
+    fSCECorrectEField(pset.get<bool>("SCECorrectEField"))
+  {
+
+    //Get the ESTAR lookup curve file name
+    std::string fname;
+    cet::search_path sp("FW_SEARCH_PATH");
+    auto ESTARPath = pset.get<std::string>("ESTARFname");
+    if (!sp.find_file(ESTARPath, fname)) {
+      throw cet::exception("ShowerESTAREnergy") << "Could not find the ESTAR lookup curve file.";
+    }
+
+    std::string ESTAR_TGraph_name = pset.get<std::string>("ESTARTGraph");
+
+    TFile fin(fname.c_str(), "READ");
+    if (!fin.IsOpen()) {
+      throw cet::exception("ShowerESTAREnergy") << "Could not read the ESTAR file. Stopping";
+    }
+
+    // Get the TGraph.
+    t_estar = dynamic_cast<TGraph2D*>(fin.Get(ESTAR_TGraph_name.c_str()));
+    if(!t_estar){
+      throw cet::exception("ShowerESTAREnergy") << "Could not read the ESTAR TGraph";
+    }
+  }
+
+  ShowerESTAREnergy::~ShowerESTAREnergy()
+  {
+  }
+
+  int ShowerESTAREnergy::CalculateElement(const art::Ptr<recob::PFParticle>& pfparticle,
+        art::Event& Event,
+        reco::shower::ShowerElementHolder& ShowerEleHolder
+  ){
+
+    if(fVerbose){
+      std::cout << "~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Shower ESTAR Energy Tool ~~~~~~~~~~~~~~~~~~~~~~~~~~~~" << std::endl;
+      if(fSCECorrectEField){std::cout << "Applying SCE correction" << std::endl;}
+    }
```

Comment:
Something to consider for later on: using the art message facility rather than in-house verbosity settings